### ISSUE TITLE
Model events fired multiple times

### ISSUE THREAD

**Issue by Fl0Cri** [wintercms/winter#1251]
### Winter CMS Build

1.2

### PHP Version

8.1

### Database engine

MySQL/MariaDB

### Issue description

Note: I will use `afterSave` as an example, but this issue applies to other events as well.

When an `afterSave()` method is defined in a model, each time the save action is run, `afterSave()` is bound to the `model.afterSave` event, and then `model.afterSave` event is fired.

Here is the code I am talking about: https://github.com/wintercms/storm/blob/54a0814cc44ac0fd2572e0da5f1dbab504979871/src/Database/Model.php#L227-L237

This means that if we call `save()` twice on the same model instance, `afterSave()` will be run once on the first time, which is expected, but twice on the second time (3 times in total).

I was about to do a PR replacing [this](https://github.com/wintercms/storm/blob/54a0814cc44ac0fd2572e0da5f1dbab504979871/src/Database/Model.php#L231) `bindEvent` by `bindEventOnce` to solve the problem, but it may not be the best option.

Using `bindEventOnce` would fix this issue in most cases, but there's one I can think of  where it would fail: if someone registers a handler for `model.afterSave` with more priority than the default, and this handler sometimes returns `false`.  
In this case, let's imagine we save the model a first time: we bind the `afterSave()` method once; we fire the event; it gets cancelled by the other handler; so we don't call the `afterSave()` method.  
Then we save a second time: we bind `afterSave()` again (we now have 2); we fire the event; for some reason the other handler does not return false; we call `afterSave()` twice.

Maybe a better approach would be to bind the `afterSave()` method to `model.afterSave` on every model instance on construction (i.e. taking it out of `self::$eventMethod(...)`), but it would require a bit more rewriting.  
As is, it won't work because of the exit condition [here](https://github.com/wintercms/storm/blob/54a0814cc44ac0fd2572e0da5f1dbab504979871/src/Database/Model.php#L212). We would need to run `bindEvent` on every model instance, but `self::$eventMethod` only once per model class.

There would be one side-effect thoug (I guess it's an undefined behavior we can change, but I mention it just in case): the `afterSave()` method would be bound early in the object construction instead of at the time the save is made. If someone registers another `model.afterSave` handler with default priority, the order in which the handlers are called may change.

Let me know what you prefer. My initial guess is that using `bindEventOnce` requires less change and would mostly fix the issue without altering existing behavior; the other solution is less hacky but requires more changes and testing. Or maybe you have other ideas ?

Let me know if you would like a PR.

### Steps to replicate

Using a model like this :
```php
class TestModel extends Model
{
    protected $table = 'test';

    public function afterSave()
    {
        dump("afterSave");
    }
}
```

Demonstration using `artisan tinker`:
```
> $test = new TestModel();
>
> $test->save();
"afterSave"
>
> $test->save();
"afterSave"
"afterSave"
>
> $test->save();
"afterSave"
"afterSave"
"afterSave"
>
```

### Workaround

**Option 1**: As only the `afterSave` model method is affected, `TestModel` can be adapted like this:
```php
class TestModel extends Model
{
    protected $table = 'test';

    public function __construct(array $attributes = [])
    {
        parent::__construct($attributes);

        $this->bindEvent('model.afterSave', function () {
            dump("afterSave");
        });
    }
}
```

Result using `artisan tinker`:
```
> $test = new TestModel();
>
> $test->save();
"afterSave"
>
> $test->save();
"afterSave"
>
> $test->save();
"afterSave"
>
```

**Option 2**: fetch a new instance of the model between saves
```
> $test = new TestModel();
>
> $test->save();
"afterSave"
> $test = TestModel::findOrFail($test->id);
>
>$test->save();
"afterSave"
> $test = TestModel::findOrFail($test->id);
>
```

**Comment by mjauvin** [comment]
I was able to reproduce.

**Comment by LukeTowers** [comment]
Issue was caused by https://github.com/wintercms/storm/commit/4f683cc3f1e001fc66c41a708a888985ef89497e / https://github.com/wintercms/storm/pull/150

**Comment by LukeTowers** [comment]
@Fl0Cri thanks very much for the detailed issue! I think probably our best option at this point is to make use of bindEventOnce() and document the potential issue that you mentioned where that listener being interrupted can result in an unintentional extra listener being left behind.

Something else we could consider is adding some form of "identified event" support to the event emitter trait so that we can explicitly unbind the generated listener after we fire the event which would solve the issue around potentially having the listeners left behind after being interrupted early. Perhaps we could do something with annonymous classes wrapping the actual handler logic so that we can either remove it ourselves from outside of the event emitter trait or otherwise prevent it from actually calling the model method if it wasn't added on that exact iteration of the listener?

At any rate we'll definitely want tests on whatever change is made :)

@bennothommo @jaxwilko any thoughts?

**Comment by bennothommo** [comment]
@Fl0Cri @LukeTowers sounds to me that the `$eventsBooted` flag is not doing its job then (https://github.com/wintercms/storm/blob/54a0814cc44ac0fd2572e0da5f1dbab504979871/src/Database/Model.php#L210-L213). That's supposed to pick up when the "nicer" events have been bound and should only do it for once per model class.

I'd dare say it's because it's using late static binding for the `$eventsBooted` property, which means it's being newly defined for each individual class, which defeats the purpose.

**Comment by LukeTowers** [comment]
@bennothommo it's because the "nicer" events (i.e. the local modal events) are actually bound inside of the listener to the laravel events (i.e. `::saving($handler)` etc). So the laravel listeners are only added once, but the laravel listeners themselves are actually binding to the local event every time they're fired.

In theory maybe we could move the binding to outside of the listener for that event, but then it might not pick up on dynamic methods added (since it's using $instance->methodExists()) after the booting of the model but before the actual listener is fired.

Let me know if you'd like to have a call to discuss

**Comment by bennothommo** [comment]
@Fl0Cri I've added a test case to Storm for this issue - changing `bindEvent` to `bindEventOnce` does fix the issue.

If you want to submit a PR for that to the [Storm library](https://github.com/wintercms/storm), feel free so you get the credit for it.

**Comment by Fl0Cri** [comment]
Ok, I've submitted the PR I prepared, replacing all instances of `bindEvent` by `bindEventOnce` inside Laravel events (from https://github.com/wintercms/storm/commit/4f683cc3f1e001fc66c41a708a888985ef89497e)

Regarding a note in the documentation, I am not sure how to formulate this, so feel free to do it if necessary ;-) (maybe it can get fixed in a near future and there's no need to document it).

As @LukeTowers mentioned, having the ability to unbind a given listener without clearing the whole event could help to solve this problem. (Like JavaScript's `removeEventListener`, but I don't think we can compare closures in PHP, so that would be hard to implement... unless the serializable closures used internally can be ?)

**Comment by LukeTowers** [comment]
@Fl0Cri the main problem with unbinding the listener is that we don't return anything useful from bindEvent. Did you have any thoughts on the idea to move the binding the of the listener outside of the laravel listener itself?

**Comment by Fl0Cri** [comment]
Sure, my initial idea of moving the event binding out of the Laravel listener would look something like this:
```php

protected function bootNicerEvents()
{
    $nicerEventsMapping = [
        'creating' => 'beforeCreate',
        'created' => 'afterCreate',
        'saving' => 'beforeSave',
        //...
    ];

    // Extend every model instance to bind method-events
    foreach ($nicerEventsMapping as $nativeEvent => $nicerEvent) {
        self::extend(function ($model) use ($nicerEvent) {
            if ($model->methodExists($nicerEvent)) {
                $model->bindEvent("model.$nicerEvent", [$model, $nicerEvent]);
            }
        });
    }

    if (isset(static::$eventsBooted[get_called_class()])) {
        return;
    }

    // Bind to Laravel events only once per class
    foreach ($nicerEventsMapping as $nativeEvent => $nicerEvent) {
        self::$nativeEvent(fn ($model) => $model->fireEvent("model.$nicerEvent", halt: true));
    }

    // TODO: booted => afterBoot, if there's a reason why it is not part of the above (it has no "before")

    static::$eventsBooted[get_called_class()] = true;
}
```

What I like about this solution is that we don't need to worry about binding / unbinding events, which should avoid any edge case.

But as you mentioned, that may be a breaking change: if someones dynamically registers (for example) a `beforeValidate` method inside or after the model `extend`, it would not be called.

However, it is actually not possible to do this for most events (such as `beforeSave`, `afterCreate`, ...), which have an empty method defined in `Winter\Storm\Database\Model`. For these, the existing methods will always be called instead of the dynamic ones.

The other side-effect of this modification is the possible change in the order in which listeners would be triggered. I can't see any case where it could be an issue, but maybe there are.

### Alternative ways
Maybe another idea to solve the issue without changing the actual behavior could be something like the following. But that would necessitate new features in the events dispatcher.

```php
// Everything stays within the Laravel event
self::$eventMethod(function ($model) use ($method) {
    if ($model->methodExists($method)) {
        // We attempt to unbind any previous method-listener, if any, before reattaching a new one
        $model->unbindEvent('model.' . $method, [$model, $method]); // This is actually not possible
        $model->bindEvent('model.' . $method, [$model, $method]);
    }

    return $model->fireEvent('model.' . $method, halt: true);
    // Unbinding right before returning would also work, but beware of potential exceptions !
});
```

This would require the ability to unbind a specific method from the emitter. Actually, the only thing `unbindEvent` can do is to clear all listeners of `'model.' $method`, which is obviously not something we want to do.

But this would require a feature that may not be possible to implement...

A variant that may have a higher chance of being implementable could be like this.
```php
self::$eventMethod(function ($model) use ($method) {
    $eventCleaner = $model->methodExists($method)
        ? $model->bindCleanableEvent('model.' . $method, [$model, $method]);
        : null;

    try {
        return $model->fireEvent('model.' . $method, halt: true);
    }
    finally { // This will always be called when returning, or if an exception is thrown
        $eventCleaner?->unbind();
        // if ($eventCleaner) $eventCleaner(); // If $eventCleaner is a cleanup function
    }
});
```

Here, `bindCleanableEvent` would return an object or anonymous function with enough knowledge of the dispatcher to be able to unbind the event (such as an index in the events array).
